#1509 Geo UI agreement level

vedran-brnjetic opened the pull request:
WIP for #1509
So far done some visuals, although I need to add correct icons to the iconset, but I picked a few from the existing ones

![image](https://user-images.githubusercontent.com/31465141/194062589-ffbf9a18-b57a-4230-8aa6-cd8c3d677759.png)

croViking committed:
added map visualizer agreement levels control component, made changes to mapmenudata to
reflect the mockup, added placeholders for all 8 data layers

croViking committed:
Cleaned up the styles for the new component

croViking committed:
Aligned the control text to the rest of the panel

croViking committed:
Added some styling to control disabled status to agreement levels

croViking committed:
Added agreement level boxes, temp icon set

vedran-brnjetic commented on src/client/pages/Geo/GeoMap/GeoMapMenuData/geoMapMenuData.tsx:
```diff
@@ -15,7 +15,7 @@ const GeoMapMenuData: React.FC = () => {
       {selectedPanel === 'data' && (
         <div>
-          <GeoMenuItem title="Map Visualizer" tabIndex={-1}>
+          <GeoMenuItem title="Forest Cover" tabIndex={-1}>
```
This title is "Forest Cover" in the mockup

Commit:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

Commit:
Sort imports

Commit:
Add functionality for loading the agreement layer

Commit:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

Commit:
Use `useCallback` for agreement level selection to benefit from caching

joonasnuutinen commented:
Implements #1740.

I hijacked the branch based on @vedran-brnjetic's work and the updated requirements in #1740.

As per the requirements, this PR implements the functionality to load the agreement layer. The UI elements are still a work in progress and will be finalized in #1741. However, I left Vedran's UI for testing purposes, as displayed below:

https://user-images.githubusercontent.com/18427577/203964824-15bd60eb-b0b4-47b1-b643-bd147dfc9ee3.mov

eraviolo commented on src/client/pages/Geo/GeoMap/GeoMapMenuData/MapVisualizerAgreementLevelsControl/mapVisualizerAgreementLevelsControl.tsx:
```diff
@@ -0,0 +1,80 @@
+   */
+  const handleAgreementLevelSelection = useCallback(
+    async (level: number): Promise<void> => {
+      const layerQuery = forestOptions.sources.map((name) => `&layer=${name}`).join('')
```
If Hansen layer is selected among the agreement layers add to the query tree cover percentage selected with 'gteHansenTreeCoverPerc'
e.g
api/geo/layers/forestAgreement/?countryIso=FIN&layer=TandemX&layer=JAXA&layer=Copernicus&layer=Hansen&gteAgreementLevel=2&gteHansenTreeCoverPerc=10

eraviolo commented:
> Implements #1740.
> 
> I hijacked the branch based on @vedran-brnjetic's work and the updated requirements in #1740.
> 
> As per the requirements, this PR implements the functionality to load the agreement layer. The UI elements are still a work in progress and will be finalized in #1741. However, I left Vedran's UI for testing purposes, as displayed below:
> 
>  Screen.Recording.2022-11-25.at.12.35.28.mov

Looks good!

minotogna reviewed (commented):
Great job ! 

However I would suggest following changes to better isolate single responsibilities and to better comply with our code standards.

### 1. Components and their sass file must be renamed with upper case convention

### 2. Create a MapController responsible for actions to be performed on google map instance.

**2.1** this file can be placed at the moment inside src/client/utils for now. we'll see if it needs better place later on

**2.2** 
```typescript
export class MapController {
  ... 
  constructor(map:google.maps.Map){
    this.map = map
  }

  addEarthEngineLayer(props:{mapId: string, mapLayerKey: string}):void {
  ...
  }

  ... other methods if needed
}
```

**2.3** this will be then used in MapVisualizerPanel.tsx
```typescript
const mapControllerRef = useRef<MapController>(new MapController(map))
```

### 3. move layers declared MapVisualizerPanel.tsx in a file in the same folder than MapVisualizerPanel.tsx + create enum LayerKey with JAXA, TandemX, Hansen...
these keys will be used in the layers object and everywhere they are now referenced as strings 

### 4. use redux actions/store to load/store layers from backend 
GeoActions.addForestLayer 
GeoActions.removeForestLayer 

### 5. in MapVisualizerPanel rename onCheckboxClick to toggleLayer. the action is important, not who trigger it.


How does this sound to you @joonasnuutinen ?

minotogna commented:
> Great job !
> 
> However I would suggest following changes to better isolate single responsibilities and to better comply with our code standards.
> 
> ### 1. Components and their sass file must be renamed with upper case convention
> ### 2. Create a MapController responsible for actions to be performed on google map instance.
> **2.1** this file can be placed at the moment inside src/client/utils for now. we'll see if it needs better place later on
> 
> **2.2**
> 
> ```ts
> export class MapController {
>   ... 
>   constructor(map:google.maps.Map){
>     this.map = map
>   }
> 
>   addEarthEngineLayer(props:{mapId: string, mapLayerKey: string}):void {
>   ...
>   }
> 
>   ... other methods if needed
> }
> ```
> 
> **2.3** this will be then used in MapVisualizerPanel.tsx
> 
> ```ts
> const mapControllerRef = useRef<MapController>(new MapController(map))
> ```
> 
> ### 3. move layers declared MapVisualizerPanel.tsx in a file in the same folder than MapVisualizerPanel.tsx + create enum LayerKey with JAXA, TandemX, Hansen...
> these keys will be used in the layers object and everywhere they are now referenced as strings
> 
> ### 4. use redux actions/store to load/store layers from backend
> GeoActions.addForestLayer GeoActions.removeForestLayer
> 
> ### 5. in MapVisualizerPanel rename onCheckboxClick to toggleLayer. the action is important, not who trigger it.
> How does this sound to you @joonasnuutinen ?

@joonasnuutinen , if you already moved to the next task, we can refactor this in a separate task.

Commit:
Change component and sass files to uppercase 1/2

An intermediate step to deal with potential inconsistencies in how Git
and GitHub handle cases

Commit:
Change component and sass files to uppercase 2/2

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

Commit:
Delegate map actions to a map controller

Commit:
Merge branch '1509-geo-ui-agreement-level' of github.com:openforis/fra-platform into 1509-geo-ui-agreement-level

Commit:
Move layers to a separate file

Commit:
Fix empty request uri

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

Commit:
Add action for getting a forest layer

Commit:
Add `addForestLayer` reducer

Commit:
Merge branch '1509-geo-ui-agreement-level' of github.com:openforis/fra-platform into 1509-geo-ui-agreement-level

Commit:
Add `removeForestLayer` reducer

Commit:
Dispatch new GeoActions

Commit:
Remove trivial type conversion

Commit:
Fix forest source not being removed

Commit:
Rename getForestLayer props type

Commit:
Use getForestLayer action

Commit:
Rename action

Commit:
Store Hansen percentage in Redux

Commit:
Remove debounce from Hansen percentage

This was needed for the slider UI but with radio buttons, not necessary.

Commit:
Use Hansen percentage directly in layer options panel

Commit:
Move `key` to correct place

Commit:
Rename component to uppercase step 1/2

An intermediate step to avoid issues with Git/GitHub case-insensitivity

Commit:
Rename component to uppercase step 2/2

Commit:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

Commit:
Cache Hansen layers based on percentage

Commit:
Avoid duplicate selected layers

Commit:
Add Hansen percentage to agreement uri

Commit:
Don't hard-code 'Hansen'

joonasnuutinen commented:
@minotogna 

1. Renamed files to uppercase. There are still other files outside the scope of this PR that have lowercase naming -> future refactoring?
2. MapController created
3. Layers moved to a separate file and typed based on `@meta/geo/forest`
4. New Redux actions created and used to abstract business logic away from components
5. Renamed action

@eraviolo

Added Hansen percentage to agreement uri.

Hansen percentage selector doesn't work correctly: sometimes you need to click the checkbox twice. However, this is outside the scope of this PR so I'll create a new issue.

joonasnuutinen committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

minotogna commented:
@joonasnuutinen can we have an up to date video?

joonasnuutinen commented:
@minotogna Sure!


https://user-images.githubusercontent.com/18427577/206452018-1a640f33-908f-4d0c-9500-9862b47e6be9.mov

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

mergify[bot] committed:
Merge branch 'master-meta' into 1509-geo-ui-agreement-level

minotogna commented on src/client/store/ui/geo/actions/getForestLayer.ts:
```diff
@@ -0,0 +1,17 @@
+  uri: string
+}
+
+export const getForestLayer = createAsyncThunk<[string, string], GetForestLayerProps>(
```
i would prefer to make return type explicit with objects. e.g. ```createAsyncThunk<{key:string,mapId:string}, GetForestLayerProps>(```

minotogna reviewed (approved):
That's a really good PR.
I left just a minor comment that could be addressed in a later pr if we want.
